### ISSUE TITLE
template + generic param + when conditional -> weird type mismatch

### ISSUE THREAD

**Issue by jangko** [nim-lang/Nim#7995]
```Nim
import typetraits

template copyRange[T](dest: seq[T], destOffset: int) =
  when supportsCopyMem(T):
    echo "A"
  else:    
    echo "B"

var a = @[1, 2, 3]
copyRange(a, 0)
```

error messages:
```text
tmpbug.nim(10, 10) template/generic instantiation from here
tmpbug.nim(4, 23) Error: type mismatch: got <int>
but expected one of:
proc supportsCopyMem(t: typedesc): bool

expression: supportsCopyMem(T)
```

the proc version works as expected

**Comment by Vindaar** [comment]
If you need a template, you can replace the `when` statement by:
```nim
when supportsCopyMem(type(T)):
```
as a workaround.
For some reason (maybe that's how generic templates work?) in case of the template `T` doesn't seem to be a normal `typedesc` variable anymore. E.g. if `T` is an `int` you can even echo it and get some random number (for floats we get an `AssertionError` though).

**Comment by jangko** [comment]
the type(T) workaround and the echo-able 'T' indicated that the 'T' is not a typedesc anymore, but a value.
and that is very weird.

**Comment by jangko** [comment]
possibly related to #5926

**Comment by metagn** [comment]
fixed by #22535
